### PR TITLE
Fix multiprocess

### PR DESCRIPTION
**Pull request by p-slash**
I wanted initiate this even though not fully tested. It also contains my requirement changes to make installation work in NERSC.
Basically two changes:
1) `Forest.mask_fields` is modified before reading files.
2) A small revamp of parallel reading using `imap` and an intermediate reader class. Each process reads and saves forest into dictionary. Then these dictionaries are merged one by one to the master dict.
Both these changes worked (ran without an error) for Pk1D + BalMask. 

Edit apr 29: I have some more changes, but don’t want to push since testing fails. Sorry didn’t know there was an automatic test!

### CODE REVIEW

**Comment by iprafols** [comment]
@corentinravoux, given that this PR is still not ready for review (I think) I marked it as draft PR for the time being. If it's ready (or when it's ready) please press the ready for review and request a reviewer

**Comment by Waelthus** [comment]
given that we'd want to merge to master and that no comments are currently here, I'm changing the base version to master for this PR

**Comment by p-slash** [comment]
I performed the data_tests on NERSC and passed it. The problem seems to be with coveralls. Not sure why, but it could be due to authentication since this from a fork. Otherwise I think the code is ready. mask_fields are set in constructor of DesiHealpix, which is inherited by DesiMocks; and strictly speaking this fix happens after all the data is read.

**Comment by Waelthus** [comment]
I re-ran the tests in case that was just a coveralls hickup, else I'll merge the fork to an igmhub branch and re-open the PR for merging that into master...

File: py/picca/delta_extraction/data_catalogues/desi_healpix.py [hunk 1]
```diff
@@ -346,8 +207,181 @@ def read_file(self, filename, catalogue, forests_by_targetid):
...
+            filename = (
+                f"{input_directory}/{healpix//100}/{healpix}/{coadd_name}-{survey}-"
+                f"dark-{healpix}.fits")
+
+            arguments.append((filename,group))
+
+        self.logger.info(f"reading data from {len(arguments)} files")
+
+        if self.num_processors>1:
+            with multiprocessing.Pool(processes=self.num_processors) as pool:
```
Comment (Waelthus) [inline py/picca/delta_extraction/data_catalogues/desi_healpix.py:363]: I think the only part that really needs to be seperated for single/multi-core processing is the `imap_it=` vs the `reader=` line, the merging could be outside the if clause (minimizing code duplication)

Comment (Waelthus) [reply py/picca/delta_extraction/data_catalogues/desi_healpix.py:363]: i.e. maybe just do `imap_it=(reader(arg) for arg in arguments)` in the sequential version and have the merging for loop outside

**Review by Waelthus — APPROVED** [review]

**Comment by Waelthus** [comment]
I did add a coulple comments here, but will merge this to an igmhub branch for tests now

File: py/picca/delta_extraction/data_catalogues/desi_healpix.py [hunk 2]
```diff
@@ -346,8 +207,181 @@ def read_file(self, filename, catalogue, forests_by_targetid):
 
                 # keep the forest
                 if targetid in forests_by_targetid:
-                    existing_forest = forests_by_targetid[targetid]
-                    existing_forest.coadd(forest)
-                    forests_by_targetid[targetid] = existing_forest
+                    forests_by_targetid[targetid].coadd(forest)
```
[the remark this answers is not in the document]
Comment (p-slash) [reply py/picca/delta_extraction/data_catalogues/desi_healpix.py:210]: My understanding is that coadding function doesn't return a forest, but modifies the parent forest. That's how I read the initial piece of code. Is it not true?

File: py/picca/delta_extraction/data_catalogues/desi_healpix.py (hunk 1 above)
Comment (p-slash) [reply py/picca/delta_extraction/data_catalogues/desi_healpix.py:363]: We can also not check for number of processors. It should behave like a serial code when `num_processors=1`. I don't know if that would create some weird behavior

Comment (Waelthus) [reply py/picca/delta_extraction/data_catalogues/desi_healpix.py:363]: the issue here is that it still would be run through multiprocessing stuff which makes debugging painful even if only on a single CPU, so having the serial case multiprocessing-free is just a convenience to allow debugging issues more easily

Comment (p-slash) [reply py/picca/delta_extraction/data_catalogues/desi_healpix.py:363]: Ok that makes sense. I'm worried about reducing the code duplication though. You can't take imap outside of `with`. For sequential version, keeping all in a list as you suggest might be memory intense. Let's postpone this to later testing if that's ok?

**Pull request merged**